006-strimzi-ui.md

```diff
@@ -0,0 +1,129 @@
+- Should be fully translatable, and accessible
+
+I would suggest that the UI implementation to follow the [Model View Controller](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) (MVC) pattern, with clear separation between business logic, and the view logic that renders the state of the UI. This would enable/allow the view logic to be modified/swapped in/out as required, but keep the same business logic. I would suggest using the [Carbon](https://www.carbondesignsystem.com/) design system as the view layer for the contributed UI, given it is open source, is supported, and I can provide both the design and implementation using it. As an example, the following is a mock up of how a Topics page could look using the Carbon design system, which would allow a user to view, create, edit and delete topics in a given cluster:
+
+![Topics mock up](./006-topicsdesign.png)
+
+By maintaining and enforcing a thin view layer, this should not stop other view layers being designed, implemented and contributed in other frameworks, such as [PatternFly](https://www.patternfly.org/v4/) or [Material](https://material.io/) for example.
+
+I would also suggest using Express as a server for this UI. This is due to its small footprint, modularity and available 3rd party modules, such as [helmet for http security](https://helmetjs.github.io/), [Passport.js for authentication and authorization support](http://www.passportjs.org/) or [the graphql and express-graphql modules for GraphQL server support](https://graphql.org/graphql-js/running-an-express-graphql-server/).
+
+This UI would also be provided with a full set of supporting elements - such as end to end tests, automation (examples of which are [in this section](#a-ui-repository-and-ways-of-working-in-the-repository)) to manage common tasks, and documentation around implementation approach, UI best practise and so on.
+
+The below shows how these pieces could integrate.
+
+![Suggested topology](./006-topology.png)
+
+Where:
+
+- 'strimzi-api' is a backend server for Kafka data/requests
+- 'Supporting elements' could be other Kafka related deployments such as Mirror Maker, Kafka Connect and Cruise control. These could also be non Kafka related deployments (not deployed as a part of Strimzi), such as LDAP servers, metric stores etc.
+
+### A UI repository, and ways of working in the repository
+
+In addition to the implementation detailed above, I would propose that any UI contributed to Strimzi be managed and developed as follows:
+
+- Be developed in a dedicated repository inside the Strimzi Github organisation
+- Both the UI server and client are contained in this repo - worked as sub modules in a monorepo
+- That design/discussion/implementation and defect issues are kept solely in this repository
+- All development is done in a behavioural manner - focusing on the end user, and the task they are trying to achieve (see [topic list](#topic-list-page) for an example)
```

Comment:
What personae are we targeting with this UI. e.g. which of the following does it cater to:

* developers of Kafka applications
* operators of Kafka applications
* operators of the Kafka cluster (and that can potentially be split into someone providing a Kafka service, and someone consuming that Kafka service who is separate from the operator for a Kafka application running on it).

Reply:
In time I think it could be all of the above. I would suggest initially we focus on the operator(s) of a Kafka cluster - covering the exact usages you mention above.

Reply:
I guess that the users of this UI should authenticate somehow. Are we going to use Kafka ACLs for that or some other system? Using ACLs provides authorization for administrative operations like altering cluster config or altering topics but it even provides auth mechanism for clients using the topics which could be anyway configured somehow using this UI (i.e. listing the topics and specifying who can write/read on topics).

Reply:
It should indeed (with the option to enable/disable it as required). I think the UI should maintain its own model of auth/authz on resources, and that this then mapped onto a backing implementation. This backing implementation could be swapped in/out as users require. What do you think of this concept, and using Kafka users (for example) to enable this for Strimzi users?